pkg/testutils/testserver.go

```diff
@@ -0,0 +1,180 @@
+	}
+
+	// Create API server client kubeconfig, used by the authenticator to update its mapping store when using CRD or EKSConfigMap
+	if err := kubeconfig.CreateAPIServerClientKubeconfig(cert, kubeAPIServerClientConfig.BearerToken, cfg.Kubeconfig, kubeAPIServerClientConfig.Host); err != nil {
+		t.Fatal(err)
+	}
+
+	printFile(cfg.Kubeconfig)
+	printFile(cfg.GenerateKubeconfigPath)
+
+	mappers, err := server.BuildMapperChain(cfg)
+	if err != nil {
+		t.Fatal(fmt.Sprintf("failed to build mapper chain: %v", err))
+	}
+	for _, m := range mappers {
+		if err := m.Start(stopCh); err != nil {
+			t.Fatal(fmt.Sprintf("start mapper %q failed: %q", m.Name(), err))
+		}
+	}
+
+	go func() {
+		httpServer := server.New(cfg, mappers)
+		httpServer.Run(stopCh)
+	}()
+
+	err = wait.PollImmediate(100*time.Millisecond, 10*time.Second, func() (done bool, err error) {
+		t.Log("Checking authenticator server health...")
+		done, err = checkHealth(cfg)
+		if err != nil {
+			t.Log(err)
+			return false, nil
+		}
+		if !done {
+			return false, nil
+		}
+		t.Log("Authenticator server healthy.")
+		return true, nil
+	})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Create aws-iam-authenticator client
+	kubeAPIServerClientConfig.ExecProvider = &clientcmdapi.ExecConfig{
+		Command:         setup.AuthenticatorClientBinaryPath,
+		Args:            []string{"token", "-i", "test-cluster"},
+		APIVersion:      "client.authentication.k8s.io/v1beta1",
+		InteractiveMode: clientcmdapi.NeverExecInteractiveMode,
+	}
+	kubeAPIServerClientConfig.BearerToken = ""
+
+	clientWithExecAuthenticator, err := client.NewForConfig(kubeAPIServerClientConfig)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	return adminClient, clientWithExecAuthenticator
+}
+
+func testConfig(t *testing.T, setup AuthenticatorTestFrameworkSetup) (config.Config, error) {
+	testDir, _ := os.MkdirTemp("", "test-integration-"+t.Name())
+
+	cfg := config.Config{
+		PartitionID:            "aws",
+		ClusterID:              "test-cluster",
+		Hostname:               "localhost",
+		HostPort:               hardcodedAuthenticatorServerPort,
+		KubeconfigPregenerated: true,
+		Address:                "127.0.0.1",
+		Kubeconfig:             filepath.Join(testDir, "apiserver.kubeconfig"),
+		GenerateKubeconfigPath: filepath.Join(testDir, "webhook.kubeconfig"),
+		BackendMode:            []string{"EKSConfigMap"},
+	}
+
+	if setup.ModifyAuthenticatorServerConfig != nil {
+		setup.ModifyAuthenticatorServerConfig(&cfg)
+	}
+
+	// validate any overrides
+	if cfg.ClusterID == "" {
+		return cfg, errors.New("cluster ID cannot be empty")
+	}
+
+	partitionKeys := []string{}
+	partitionMap := map[string]endpoints.Partition{}
+	for _, p := range endpoints.DefaultPartitions() {
+		partitionMap[p.ID()] = p
+		partitionKeys = append(partitionKeys, p.ID())
+	}
+	if _, ok := partitionMap[cfg.PartitionID]; !ok {
+		return cfg, errors.New("Invalid partition")
+	}
+
+	if errs := mapper.ValidateBackendMode(cfg.BackendMode); len(errs) > 0 {
```

Comment:
is this and checking cfg.ClusterID necessary, shouldn't the server do these checks for us when you start it? when you call mapper.Start it should crap out if you provide an invalid one right?

Reply:
It calls `ValidateBackendMode()` in `cmd/aws-iam-authenticator/root.go`, so while I'm not sure if its strictly necessary for a test server, I'm just copying the `getConfig()` function there.

Reply:
I would like to move `BuildMapperChain()` from cmd to pkg, as I'm not using cmd in the integration tests at this point and I'd rather not.  Maybe I can do the same for `ValidateBackendMode()`.